### ISSUE TITLE
Constant from extension method?

### ISSUE THREAD

**Issue by mnordine** [dart-lang/language#663]
Any way something like this could work?

```dart
extension Time on int {
  Duration get ms => 
    Duration(microseconds: (this * Duration.microsecondsPerMillisecond).toInt());
}

main() {
  // const time = Duration(milliseconds: 250);
  const time = 250.ms;
  print(time);
}
```

as long as `this` is itself `const`?

**Comment by eernstg** [comment]
It is definitely pretty far away from the constants that we support today: Extension methods can never be part of a constant expression evaluation (that would enable constant evaluation to run arbitrary code), and the only non-operator member that we can evaluate in a constant is `length`, and that's only with an instance of `String`.

We have many requests for more powerful constant evaluation, but it is a very delicate area (there is no end to the complexities popping up when we consider constant evaluation together with `fromEnvironment`, modular compilation, and whatnot).

Granted&mdash;it looks nice! But there are lots of reasons why it won't be easy to make it happen..

(OK, here's a loophole: if we consider a compilation process where `.ms` is inlined prior to constant evaluation then it turns out to be an OK constant expression. But we don't have any such loophole, and I have no idea how many implications it would have to try to do such a thing. Maybe we could call this loophole "macros". But we don't have those, and it's definitely not a trivial thing to add. ;-)

**Comment by lrhn** [comment]
That "loophole" would be equivalent to having "potentially constant methods", methods which are acceptable in constant expressions because they only depend on `this` and other potentially constant expressions. You would have to declare such a method, so that we can check it, and it likely can't be virtual (which means it's a slam-dunk fit for extension methods). Perhaps:
```dart
class C {
   final int x;
   const C(this.x);
   const int next => x + 1;
}
````
Then `const c0 = C(0); const c1 = C(c0.next);` would b allowed because `c0.next` is a (potentially) `const` method on a const object.

That's a rather large increase in complexity and expressive power for const expressions, and we have no current plans to do add any new major features to constant expressions.

**Comment by eernstg** [comment]
Cf. https://github.com/dart-lang/sdk/issues/29277, https://github.com/dart-lang/sdk/issues/20962 as well.

**Comment by rmortes** [comment]
Sorry for commenting on such an old issue, but is this feature being considered at the time?

**Comment by lrhn** [comment]
No. There is no current work being done on allowing potentially constant functions.
